File: java/src/main/java/com/packman/Pacman.java
package com.packman;

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Entity Class — Pacman as a Game Object
// ═══════════════════════════════════════════════════════
// Pacman is a self-contained entity with position (gridX, gridY),
// movement direction (dirX, dirY), and buffered input (nextDirX,
// nextDirY). It knows how to update itself but does NOT draw itself.
// Rendering is handled separately by GamePanel (separation of concerns).
//
// Compare with:
//   Python (Pygame) → class Pacman(pygame.sprite.Sprite): with
//                      self.rect for position, self.image for display
//   JavaScript (Canvas) → Plain object { x, y, dx, dy } with
//                          external renderer drawing it
//   TypeScript → class Pacman with typed properties: x: number, etc.
//   C++ (SDL2) → struct Pacman { int x, y; SDL_Texture* tex; }
//
// KEY CONCEPT: Separating entity state from rendering logic is a
// fundamental software design principle. The entity class holds data
// and game logic; a dedicated renderer reads that data to draw.
// This makes entities testable without a graphics context and allows
// swapping renderers (e.g., Swing to JavaFX) without touching entities.
// ═══════════════════════════════════════════════════════
public class Pacman {
    private int gridX;
    private int gridY;
    private int size;
    private int dirX = 0;
    private int dirY = 0;
    private int nextDirX = 0;
    private int nextDirY = 0;

    public Pacman(int x, int y, int size) {
        this.gridX = x;
        this.gridY = y;
        this.size = size;
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Input Buffering — Direction Queuing
    // ═══════════════════════════════════════════════════════
    // setDirection() stores the DESIRED direction, not the active
    // one. The actual direction change happens in update() only
    // if the new direction is valid (no wall). This means:
    //   - Player presses LEFT near a junction
    //   - nextDirX/Y is set to (-1, 0) immediately
    //   - On next update(), if left is open, direction changes
    //   - If left is a wall, Pacman keeps moving in current dir
    //
    // Compare with:
    //   Python (Pygame) → Store next_direction in event handler,
    //                      apply in update() after wall check
    //   JavaScript (Canvas) → Same pattern; nextDir set in keydown,
    //                          validated in update()
    //   TypeScript → Same as JS with typed direction enum/tuple
    //   C++ (SDL2) → Same pattern; queue direction in event handler
    //
    // KEY CONCEPT: Input buffering makes grid-based movement feel
    // smooth. Without it, pressing a key one frame too early at a
    // junction would be ignored, frustrating the player.
    // ═══════════════════════════════════════════════════════
    public void setDirection(int x, int y) {
        nextDirX = x;
        nextDirY = y;
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Object Movement — Grid-Based with Wall Checking
    // ═══════════════════════════════════════════════════════
    // Movement happens one tile per update. The algorithm:
    //   1. Try the buffered direction (player's latest input)
    //   2. If that's blocked by a wall, try continuing in current direction
    //   3. If both are blocked, Pacman stands still
    //
    // This creates the classic Pacman feel: you can "pre-turn" by
    // pressing a direction before reaching the junction.
    //
    // Compare with:
    //   Python (Pygame) → rect.x += speed * dx; then check colliderect()
    //                      with walls; undo if colliding
    //   JavaScript (Canvas) → x += speed * dx; check collision; revert
    //   TypeScript → Same as JS with typed position/velocity
    //   C++ (SDL2) → Similar check-then-move or move-then-revert pattern
    //
    // KEY CONCEPT: Grid-based movement simplifies collision — you check
    // BEFORE moving, not after. Pixel-based movement often uses
    // "move then correct" (push back if overlapping). Grid-based
    // is simpler but limits movement to discrete tile steps.
    // ═══════════════════════════════════════════════════════
    public void update(Maze maze) {
        // Try to move in the next direction if possible
        int newX = gridX + nextDirX;
        int newY = gridY + nextDirY;

        if (isValidMove(newX, newY, maze)) {
            dirX = nextDirX;
            dirY = nextDirY;
            gridX = newX;
            gridY = newY;
        } else {
            // Try to continue in current direction
            newX = gridX + dirX;
            newY = gridY + dirY;
            if (isValidMove(newX, newY, maze)) {
                gridX = newX;
                gridY = newY;
            }
        }
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Wall Collision — Boundary Checking
    // ═══════════════════════════════════════════════════════
    // isValidMove delegates to Maze.isWall(), which handles
    // bounds checking. This is a "look before you leap" pattern:
    // check the destination BEFORE moving there.
    //
    // Compare with:
    //   Python (Pygame) → if not maze[new_y][new_x].is_wall: move
    //   JavaScript (Canvas) → if (!maze.isWall(newX, newY)) { ... }
    //   TypeScript → Same as JS; maze.isWall(x: number, y: number): boolean
    //   C++ (SDL2) → if (!maze.isWall(newX, newY)) { pos = newPos; }
    //
    // KEY CONCEPT: Validation before mutation is a defensive programming
    // pattern. In games, it prevents entities from entering invalid
    // states (inside walls, out of bounds).
    // ═══════════════════════════════════════════════════════
    private boolean isValidMove(int x, int y, Maze maze) {
        return !maze.isWall(x, y);
    }

    public int getSize() {
        return size;
    }

    public int getGridX() {
        return gridX;
    }

    public int getGridY() {
        return gridY;
    }
}

File: website/frontend/public/game/js/pacman.js
// Source of truth: shared/game_constants.json
// Pure logic module — ZERO rendering/GUI code.

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Input Handling
// ═══════════════════════════════════════════════════════
// Pacman uses an "input buffering" pattern: setDirection() stores the
// DESIRED direction, but the move only executes in update() if the
// target tile is not a wall. This decouples input capture from movement
// validation. In JavaScript, keydown events fire asynchronously via the
// browser's event loop, so buffering prevents lost inputs between frames.
//
// Compare with:
//   Java → Uses a Swing KeyListener (keyPressed) to buffer direction in
//          the same way; Java's EDT (Event Dispatch Thread) serializes
//          input events, similar to the browser's single-threaded model.
//   Python → Uses pygame.event.get() polled each frame in the game loop;
//            direction is buffered identically, but input is pulled (polled)
//            rather than pushed (event-driven) as in JS and Java.
//
// KEY CONCEPT: JavaScript's event-driven architecture means keydown fires
// between animation frames. Buffering ensures the latest input is always
// available when the next update() tick runs.
// ═══════════════════════════════════════════════════════

const PACMAN_START_X = 5;
const PACMAN_START_Y = 5;

class Pacman {
    constructor() {
        this._gridX = PACMAN_START_X;
        this._gridY = PACMAN_START_Y;
        this._dirX = 0;
        this._dirY = 0;
        this._nextDirX = 0;
        this._nextDirY = 0;
    }

    /**
     * Buffers the next desired direction. The actual direction change
     * happens in update() only if the new direction is valid (no wall).
     * This input buffering makes grid-based movement feel responsive.
     */
    setDirection(dx, dy) {
        this._nextDirX = dx;
        this._nextDirY = dy;
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Object Movement
    // ═══════════════════════════════════════════════════════
    // Pacman moves exactly one grid tile per update tick. The algorithm
    // tries the buffered direction first, then falls back to the current
    // direction, then stops. This "try-preferred, fallback-to-current"
    // pattern is standard for grid-based Pac-Man movement across all
    // three language implementations.
    //
    // Compare with:
    //   Java → Identical logic in Pacman.update(Maze); Java's strong
    //          typing requires explicit int parameters but the algorithm
    //          is the same step-for-step.
    //   Python → Same algorithm in pacman.update(maze); Python uses
    //            tuples for direction (dx, dy) instead of separate fields,
    //            which is more Pythonic but functionally identical.
    //
    // KEY CONCEPT: Grid-based movement avoids floating-point issues
    // entirely. Position is always an integer tile coordinate, making
    // wall checks and collision detection trivially exact.
    // ═══════════════════════════════════════════════════════

    /**
     * Attempts to move Pacman one tile per update:
     *   1. Try the buffered direction (player's latest input)
     *   2. If blocked by a wall, try continuing in the current direction
     *   3. If both are blocked, Pacman stands still
     */
    update(maze) {
        // Try to move in the buffered (next) direction
        let newX = this._gridX + this._nextDirX;
        let newY = this._gridY + this._nextDirY;

        if (!maze.isWall(newX, newY)) {
            this._dirX = this._nextDirX;
            this._dirY = this._nextDirY;
            this._gridX = newX;
            this._gridY = newY;
        } else {
            // Fall back to continuing in current direction
            newX = this._gridX + this._dirX;
            newY = this._gridY + this._dirY;
            if (!maze.isWall(newX, newY)) {
                this._gridX = newX;
                this._gridY = newY;
            }
        }
    }

    get gridX() {
        return this._gridX;
    }

    get gridY() {
        return this._gridY;
    }
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = { Pacman };
} else if (typeof window !== 'undefined') {
    window.Pacman = Pacman;
}
